### DISCUSSION TITLE
Assigning Responsibility for Creating Response Objects: Controller versus Service Layer

### DISCUSSION THREAD

**Discussion by soobinJung** [soobinJung/JCK-World#21] [Architecture and Design] [answered]
안녕하세요,

현재 Spring Boot를 이용한 REST API 개발 중에 있습니다. 개발 과정에서 응답 객체 ResExaminationQuestion을 Controller 내에서 반환하는 것과 Service 계층에서 생성하는 것 사이의 결정에 대해 고민하고 있습니다. 구체적인 구현 방안에 대한 피드백을 구하고자 합니다.

##  Controller 에서의 생성 후 반환

ExaminationQuestionController는 ExaminationQuestionService 및 ExaminationService를 주입 받아, 주어진 examinationId에 해당하는 시험 정보와 질문들을 조회하여 ResExaminationQuestion 객체를 반환합니다. 이는 RESTful 원칙에 따라 클라이언트에게 JSON 형식으로 데이터를 전달하는 역할을 합니다.

```
@RequiredArgsConstructor
@RestController
public class ExaminationQuestionController {

    private final ExaminationQuestionService examinationQuestionService;
    private final ExaminationService examinationService;

    @GetMapping("/examination-question/{examinationId}")
    public ResExaminationQuestion findAllByExaminationId(@PathVariable Long examinationId) {
        Examination examination = examinationService.getExaminationById(examinationId);
        List<ExaminationQuestion> examinationQuestions = examinationQuestionService.findAllByExaminationId(examinationId);

        return new ResExaminationQuestion(examination, examinationQuestions);
    }
}
```

## Service 계층에서의 생성

서비스 계층의 ExaminationQuestionService에서는 ExaminationRepository와 ExaminationService를 활용하여 필요한 시험 정보와 질문들을 집합한 후, ResExaminationQuestion 객체를 구성하여 반환합니다. 이 방식은 객체 생성의 책임을 서비스 계층으로 이동시키는 것으로, 비즈니스 로직과 데이터 처리를 서비스 계층에 중앙집중화시키는 장점이 있습니다.

```
@RequiredArgsConstructor
@RestController
public class ExaminationQuestionController {

    private final ExaminationQuestionService examinationQuestionService; 

    @GetMapping("/examination-question/{examinationId}")
    public ResExaminationQuestion findAllByExaminationId(@PathVariable Long examinationId) {
        return examinationQuestionService.findAllByExaminationId(examinationId);
    }
}
```

```
@RequiredArgsConstructor
@Service
public class ExaminationQuestionService {

    private final ExaminationQuestionRepository examinationQuestionRepository;
    private final ExaminationService examinationService;

    public ResExaminationQuestion findAllByExaminationId(Long examinationId) {
        Examination examination = examinationService.getExaminationById(examinationId);
        List<ExaminationQuestion> examinationQuestions = examinationQuestionRepository.findAllByExaminationId(examinationId);

        return new ResExaminationQuestion(examination, examinationQuestions);
    }
}

```

**Comment by soobinJung** [comment]
### 1. 책임의 분리(Separation of Concerns)
##### 컨트롤러(Controller)
HTTP 요청을 처리하고 클라이언트에게 HTTP 응답을 반환하는 역할에 집중하는 것이 좋습니다. 이는 주로 요청의 유효성 검사, 요청을 처리하기 위한 서비스 메서드 호출, 그리고 처리 결과를 클라이언트에 반환하는 형태의 데이터로 변환하는 작업을 포함합니다.
##### 서비스(Service)
비즈니스 로직의 실행, 데이터의 처리 및 변환 등의 역할을 수행합니다. 서비스 계층에서 응답 객체를 생성하게 되면, 비즈니스 로직과 데이터 변환 로직을 더 가까이 배치할 수 있으며, 재사용성과 테스트 용이성 측면에서 이점을 얻을 수 있습니다.

### 2. 재사용성(Reusability)과 테스트 용이성(Testability)
응답 객체의 생성을 서비스 계층에서 처리하면, 해당 로직을 여러 컨트롤러에서 재사용할 수 있습니다. 또한, 비즈니스 로직과 밀접하게 연관된 응답 객체의 생성 과정을 서비스 계층에서 담당하게 함으로써, 단위 테스트가 용이해집니다. 이는 비즈니스 로직과 데이터 변환 로직을 함께 테스트할 수 있기 때문입니다.

### 3. 유지보수성(Maintainability)
응답 객체를 서비스 계층에서 생성하게 되면, 비즈니스 로직의 변경이 필요할 경우 해당 로직과 관련된 응답 변환 로직을 서비스 계층 내에서 모두 처리할 수 있어, 유지보수성이 향상됩니다. 컨트롤러와 서비스 계층의 역할이 명확히 분리되어 있으면, 시스템의 다양한 부분을 독립적으로 개발하고 수정할 수 있어 전체적인 시스템의 복잡성을 관리하기가 더 용이해집니다.

## 결론
응답 객체 생성의 적절한 위치는 애플리케이션의 구조와 특정 요구 사항에 따라 다릅니다. 하지만 일반적인 관행과 책임의 분리 원칙을 따른다면, 서비스 계층에서 응답 객체를 생성하는 것이 더 바람직할 수 있습니다. 이렇게 하면 컨트롤러 계층을 더 간결하게 유지하고, 비즈니스 로직과 데이터 변환 로직을 더 밀접하게 연결하여 재사용성과 테스트 용이성을 향상시킬 수 있습니다.

❓
하지만 서비스의 메소드들이 도메인 객체가 아닌 응답 전용 객체를 반환하게 되면, 다른 서비스들이 이러한 응답 객체를 필요로 할 때 문제가 발생할 수 있습니다. 예를 들어, 한 서비스가 다른 서비스를 참조하여 그 결과를 받아야 할 경우, 응답 전용 객체는 재사용성과 테스트 용이성을 저해하며, 서비스 간의 결합도를 높이는 원인이 될 수 있습니다. 따라서 서비스 메소드가 반환하는 객체는 가능한 한 재사용 가능하고, 다양한 상황에서 유연하게 사용될 수 있는 형태여야 합니다. 서비스 계층에서 비즈니스 로직을 처리한 후, 컨트롤러 계층에서 이를 적절한 응답 형태로 변환하는 것이 더 깔끔한 분리를 제공하고, 각 계층의 책임을 명확히 할 수 있다는 생각도 드는 것 같습니다.

**Comment by RyooChan** [comment, marked as answer]
@soobinJung 
뭔가 이전에 고민했던 내용과 비슷한 논의인것 같아 저도 댓글 달아봅니다!
Response 용 Object를 service단이나 controller단 둘중 어디서 다는것이 좋을까에 관한 글이 맞을까요??

- 결론부터 말하자면 저는 **둘다 각각 따로 만드는게 맞다고는 하는데, 나라면 Service에서 변환한다** 라는 입장입니다.

이유를 간단하게 적으면

- Service : 비즈니스 로직이 실제로 처리된다. 복잡성에 따라 다르지만, 여러 객체를 검색해서 필요한 데이터만 합치는 로직이 일어날 수 있다.
- Controller : 유저나 클라이언트에게 보여지는 데이터를 들고와서, 이를 return시킨다.

그러면 만약에 Controller에서 DTO를 만들게 된다면 어떤 일이 발생할지를 생각해보면

- Entity가 변경되는 경우 Controller에서도 데이터를 변경해야 하는 경우가 생긴다.
- 만약 하나의 Entity가 아니라 여러 Entity를 들고와서 이를 합친 DTO를 만든다면? Controller에서 많은 Service를 불러와야 할 것이다.

이를 Service에서 처리한다면?

- Entity가 변경되는 경우 Service에서만 처리하고 Controller에서 변경이 필요하지 않다. -> 어차피 변경해야하는게 아닐까? 하는 생각이 들 수 있지만 보통 하나의 Service를 다른 Service에서 활용할 수 있기 때문에 Controller에서 DTO를 변경하는 경우 Service / Controller 모두 변경이 필요할 수 있다는 생각이 들었습니다.
- 여러 Entity를 통한 DTO 생성의 경우 facade를 통해 만들수도 있고 Service method를 하나 더 만들면 쉽게 해결된다.

이에요!
물론 Controller, Service 모든 곳에서 DTO를 만드는게 더 좋다고는 하지만 저라면 Service에서 만드는걸 더 선호하기는 합니다!

**Reply by soobinJung** [reply]
새로운 아키텍쳐네요 !!! 이런식으로 접근하는 것도 하나의 방법이 될 수 있다는 생각이 드네요 !

계층별 dto 를 생성하는 것에 대한 개발방식은 들어봤으나 아직 실제로 해본적은 없거든요 ㅠㅠ
저도 이런식으로 한번 접근해봐야겠습니다.

좋은 의견 감사합니다 !!
